click/_termui_impl.py

```diff
@@ -108,6 +108,20 @@ def __iter__(self):
         self.render_progress()
         return self.generator()
 
+    def next(self):
+        """py2/3 compatible iterator"""
+        return self.__next__()
+
+    def __next__(self):
+        """
+        iteration is defined in terms of a generator function, returned by
+        iter(self); use that to define next()
+
+        Note that this works because `self.iter` is an iterable consumed by
+        that generator, so it is re-entry safe.
+        """
+        return next(iter(self))
```

Comment:
it should be noted that this only works, because the "generator" is an actual generator not really doing anything more than iterating one at a time

should the generator ever need to do more than basic iteration and render calls, those methods would toughly break

Reply:
Can you give an example of the kind of future behaviors which would break this?
I'm not able to think of changes there which would break this but not also break `for x in progressbar(...)` usage.

But I'm perfectly happy to try to rework this a bit if I have a clear idea of what it ought to be able to handle.

Reply:
@sirosen this currently is mostly a hypothetical concern, but we ought to keep track of it

the generator object that makes up the iterator is a state tracking object, in any case where it grows a extra behavior, its very likely that not ensuring we only have one of it around it will start behaving strange

so it should be noted at least (so if it ever happens we can blame it on not reading the comments

Reply:
Ah, I see what you're saying. I'll add it to the docstring here, since I already have the note on it being "re-entry safe" which is basically talking about the same thing.

Reply:
I've tried to clarify without bloating the docstrings too much. Hopefully the revision makes it clear that this relies on `self.generator()` being safe to call repeatedly to "resume".

Reply:
looks good to me, thanks

Reply:
Thanks for pointing this out, by the way. The prior version was _definitely_ unclear.